File: src/main/java/com/davidbarsky/dag/GraphGenerator.java
package com.davidbarsky.dag;
import java.util.ArrayList;
import java.util.Collection;
import java.util.EnumMap;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.IntStream;

import com.davidbarsky.dag.models.Task;
import com.davidbarsky.dag.models.TaskQueue;
import com.davidbarsky.dag.models.states.MachineType;
import info.rmarcus.ggen4j.GGen;
import info.rmarcus.ggen4j.GGenCommand;
import info.rmarcus.ggen4j.GGenException;
import info.rmarcus.ggen4j.graph.GGenGraph;
import info.rmarcus.ggen4j.graph.Vertex;

public class GraphGenerator {
	private GraphGenerator() { }
	
	public static List<TaskQueue> randomGraph(int numQueues) {
		List<TaskQueue> queues = new ArrayList<>(numQueues);
		Map<Integer, Task> tasks = new HashMap<>();
		int i = 0;

		IntStream.range(0, numQueues)
		.forEach(k -> queues.add(new TaskQueue(MachineType.SMALL)));

		Collection<Vertex> vertices = getErdosGNMSources();
		// round-robin each vertex to a task queue
		for (Vertex v : vertices) {
			int vertexLatency = Integer.parseInt(v.getVertexProperties().get("latency"));

			TaskQueue tq = queues.get(i);
			i = (i + 1) % numQueues;

			Map<MachineType, Integer> latencies = new EnumMap<>(MachineType.class);
			latencies.put(MachineType.SMALL, vertexLatency);

			Task t = new Task(v.getID(), tq, latencies);
			tasks.put(v.getID(), t);
			tq.add(t);
		}
		
		// add dependencies
		for (Vertex v : vertices) {
			Task t = tasks.get(v.getID());
			for (Map.Entry<Vertex, Map<String, String>> child : v.getChildren().entrySet()) {
				int childID = child.getKey().getID();
				Task dependent = tasks.get(childID);
				int networkCost = Integer.parseInt(child.getValue().get("networking"));
				
				dependent.addDependency(networkCost, t);
			}
		}

		return queues;
	}

	protected static Collection<Vertex> getErdosGNMSources() {
		GGenCommand.GGEN_PATH = System.getenv("GGEN_PATH");
		if (GGenCommand.GGEN_PATH == null) {
			throw new DAGException("You need to set the GGEN_PATH environmental variable!");
		}

		GGenGraph graph;
		try {
			graph = GGen.generateGraph().erdosGNM(20, 100)
					.vertexProperty("latency").uniform(10, 30)
					.edgeProperty("networking").uniform(5, 20)
					.generateGraph();

			return graph.allVertices();
		} catch (GGenException e) {
			throw new DAGException(e.getMessage());
		}
	}
}

File: src/main/java/sirens/experiments/GraphGenerator.scala
package sirens.experiments

import java.util

import sirens.models.Task

import collection.JavaConverters._
import info.rmarcus.ggen4j.GGen
import sirens.dag.TopologicalSorter
import sirens.models.data.{LatencyBounds, NetworkingBounds}

// @formatter:off
object GraphGenerator {
  val latencyBounds: LatencyBounds = LatencyBounds(10, 60)
  val networkingBounds: NetworkingBounds = NetworkingBounds(10, 60)

  def genericGraph(graphSize: Int): util.List[Task] = {
    val graph = GGen.generateGraph().erdosGNM(graphSize, 100)
      .vertexProperty("latency").uniform(10, 60)
      .edgeProperty("networking").uniform(10, 60)
      .generateGraph().topoSort()
    TopologicalSorter.mapToTaskList(graph.allVertices())
  }

  def cholesky(latencyBounds: LatencyBounds, networkingBounds: NetworkingBounds): util.List[util.List[Task]] = {
    (10 :: 12 :: 14 :: 16 :: 18 :: 20 :: Nil).map { n =>
      val graph = GGen.dataflowGraph().cholesky(n)
        .vertexProperty("latency").uniform(latencyBounds.lower, latencyBounds.upper)
        .edgeProperty("networking").uniform(networkingBounds.lower, networkingBounds.upper)
        .generateGraph().topoSort()

      TopologicalSorter.mapToTaskList(graph.allVertices())
    }.asJava
  }

  def fibonacci(latencyBounds: LatencyBounds, networkingBounds: NetworkingBounds): util.List[util.List[Task]] = {
    (5 :: 6 :: 7 :: 8 :: 9 :: 10 :: Nil).map { n =>
      val graph = GGen.staticGraph().fibonacci(n, 1)
        .vertexProperty("latency").uniform(latencyBounds.lower, latencyBounds.upper)
        .edgeProperty("networking").uniform(networkingBounds.lower, networkingBounds.upper)
        .generateGraph().topoSort().allVertices()

      TopologicalSorter.mapToTaskList(graph)
    }.asJava
  }

  def forkJoin(latencyBounds: LatencyBounds, networkingBounds: NetworkingBounds): util.List[util.List[Task]] = {
    (2 :: 3 ::4 :: 5 ::6 ::7 :: Nil).map { n =>
      // `15` is the diameter of the graph
      val graph = GGen.staticGraph().forkJoin(n, 15)
        .vertexProperty("latency").uniform(latencyBounds.lower, latencyBounds.upper)
        .edgeProperty("networking").uniform(networkingBounds.upper, networkingBounds.upper)
        .generateGraph().topoSort()

      TopologicalSorter.mapToTaskList(graph.allVertices())
    }.asJava
  }

  def poisson(latencyBounds: LatencyBounds, networkingBounds: NetworkingBounds): util.List[util.List[Task]] = {
    (8 :: 9 :: 10 :: 11 :: 12 :: 13 :: Nil).map { n =>
      val graph = GGen.dataflowGraph().poisson2D(20, n)
        .vertexProperty("latency").uniform(latencyBounds.lower, latencyBounds.upper)
        .edgeProperty("networking").uniform(networkingBounds.lower, networkingBounds.upper)
        .generateGraph().topoSort()

      TopologicalSorter.mapToTaskList(graph.allVertices())
    }.asJava
  }

  def sparseLU(latencyBounds: LatencyBounds, networkingBounds: NetworkingBounds): util.List[util.List[Task]] = {
    (7 :: 8 :: 9 :: 10 :: 11 :: 12 :: Nil).map { n =>
      val graph = GGen.dataflowGraph().sparseLU(n)
        .vertexProperty("latency").uniform(latencyBounds.lower, latencyBounds.upper)
        .edgeProperty("networking").uniform(networkingBounds.lower, networkingBounds.upper)
        .generateGraph().topoSort()

      TopologicalSorter.mapToTaskList(graph.allVertices())
    }.asJava
  }

  def erdos(latencyBounds: LatencyBounds, networkingBounds: NetworkingBounds): util.List[util.List[Task]] = {
    (220 :: 364 :: 560 :: 816 :: 1140 :: 1540 :: Nil).map { n =>
      val graph = GGen.generateGraph().erdosGNM(n, 100)
        .vertexProperty("latency").uniform(latencyBounds.lower, latencyBounds.upper)
        .edgeProperty("networking").uniform(networkingBounds.lower, networkingBounds.upper)
        .generateGraph().topoSort()
      TopologicalSorter.mapToTaskList(graph.allVertices())
    }.asJava
  }

  def erdosGNP(latencyBounds: LatencyBounds, networkingBounds: NetworkingBounds): util.List[util.List[Task]] = {
    (220 :: 364 :: 560 :: 816 :: 1140 :: 1540 :: Nil).map { n =>
      val graph = GGen.generateGraph().erdosGNP(n, 0.5)
        .vertexProperty("latency").uniform(latencyBounds.lower, latencyBounds.upper)
        .edgeProperty("networking").uniform(networkingBounds.lower, networkingBounds.upper)
        .generateGraph().topoSort()
      TopologicalSorter.mapToTaskList(graph.allVertices())
    }.asJava
  }
}
